xds/internal/xdsclient/tests/eds_watchers_test.go

```diff
@@ -52,6 +52,31 @@ const (
 	edsPort3 = 3
 )
 
+type endpointsUpdateErrTuple struct {
+	update xdsresource.EndpointsUpdate
+	err    error
+}
+
+type endpointsWatcher struct {
+	updateCh *testutils.Channel
+}
+
+func newEndpointsWatcher() *endpointsWatcher {
+	return &endpointsWatcher{updateCh: testutils.NewChannel()}
+}
+
+func (ew *endpointsWatcher) OnUpdate(update *xdsresource.EndpointsResourceData) {
+	ew.updateCh.Send(endpointsUpdateErrTuple{update: update.Resource})
+}
+
+func (ew *endpointsWatcher) OnError(err error) {
+	ew.updateCh.SendOrFail(endpointsUpdateErrTuple{err: err})
+}
+
+func (ew *endpointsWatcher) OnResourceDoesNotExist() {
+	ew.updateCh.SendOrFail(endpointsUpdateErrTuple{err: xdsresource.NewErrorf(xdsresource.ErrorTypeResourceNotFound, "Endpoints not found in received response")})
```

Comment:
In the code you refactored into this, any send of this tuple type is a blocking send, whether it got an Update or it got an Error. Is there a reason you switched this to a non blocking, and if none, can you make this Send blocking for verification purposes? Seems less racy?

Reply:
The earlier code was not using the go-control-plane, and the new tests use it. The go-control-plane server has this weird behavior where it keeps resending a NACKed response. This means that in cases where we deal with an error, we will keep hitting the same error again and again in our e2e tests. If I don't have this non blocking send, the test would leak this goroutine.

I remember filing an issue with the go-control-plane about this, but there was no response.

Reply:
Ah, ok, sounds good.